### ISSUE TITLE
New Doc Content Request - Detecting Omitted Arguments

### ISSUE THREAD

**Issue by librasteve** [Raku/doc#4493]
## Problem or new feature

A recent conversation on the raku-beginner channel [here](https://discord.com/channels/538407879980482560/768511641758466088/1251731925341573242)
asks 

> 03:55]pelevesque: Hey guys, I'm stumped
> 
> sub a ($scalar?) { say $scalar.defined }
> a(1); # True
> a(); # False
> 
> This works, but I need it for a list....
> 
> sub a (@list?) { say @list.defined }
> 
> That fails... 
> 
> Is there a way to know if a list as an optional parameter was defined? It seems like it creates an empty list.
> 
> A clear and concise description of what the bug is, including links to the page where you have found it.

## Suggestions

Then many ideas were proposed, but, at the end, the question remains _Is there a way to know if a list as an optional parameter was defined?_ and I think that we could improve the docs with a good, standard example of this to add at the bottom of this section: https://docs.raku.org/language/signatures#Optional_and_mandatory_arguments

My "starter for 10" is:

Checking for the presence of an optional argument is complicated by the fact that the empty List ```()```, Array ```[]``` and Hash ```%{}``` are ```.defined```. It can be done like this for Scalar parameters:

```
sub is-scalar($x? = Nil) {
    with $x {say True}      #with uses .defined
    else    {say False}
}
is-scalar(42);  #True
is-scalar();    #False
```

For Positional parameters, you can introduce a constant ```ArrayAt``` for a specific Array, and use it like this:

```
constant ArrayAt = []; # a specific Array

sub is-positional(@a? = ArrayAt) {
    if @a =:= ArrayAt {say False}
    else              {say True}
}
is-positional([^42]);   #True
is-positional(());      #True 
is-positional();        #False
```

And similarly, for Associative parameters with ```HashAt``` for a specific Hash:

```
constant HashAt = {}; # a specific Hash

sub is-associative(%h? = HashAt) {
    if %h =:= HashAt {say False}
    else             {say True}
}
is-associative({a=>1}); #True
is-associative({});    #True
is-associative();       #False
```

Note: you can use the tests ```[] ~~ Empty``` (True) and ```{} ~~ Empty``` (True) to test for the presence of elements within Positional and Associative parameters respectively.

**Comment by gfldex** [comment]
Surpsisingly, we a default undefined value can be bound to a positional argument.

```
    constant NaA = Positional; # This is Not an Array.
    sub nice(@a? = NaA) { say [@a =:= NaA, @a.defined, @a.so]; }
    nice(); # [True False False]
    nice([]); # [False True False]
```

**Comment by gfldex** [comment]
Same for Associative.

```
    constant NaH = Associative; # This is Not a Hash.
    sub nice(%h? = NaH) { say [%h =:= NaH, %h.defined, %h.so]; }
    nice();
    nice({});
```

**Comment by raiph** [comment]
I was just exploring the same thing. I'd written this so far:

Maybe:
```
sub a (@list = Positional ) { say @list.defined }
a # False
```
But then I was writing about maybe making a related change for `6.f` (`@?` means `@ = Positional` etc.), and then began exploring some related destructuring results I found surprising, and... then your comment appeared.

**Comment by raiph** [comment]
@librasteve 

```
sub a (Positional $list?) { say $list.defined }
```
I love that. But it changes the sigil, which might be an issue.

> I also find myself having to rethink my idea of an optional parameter `?` since it seems that a vanilla positional with a default is effectively optional in that the argument may be missing.

What's going on applies to _any_ parameter (positional or non-positional, scalar or non-scalar, vanilla or fancy).

Either `is optional` (or `?`) or a default (`= default`) direct Raku to bind a signature even if, for a parameter marked one of those ways, there is no corresponding argument.

(Indeed Raku(do) happily/silently lets you write both `is optional` (or `?`) _and_ `= default` if you want.)

If a signature successfully binds, then Raku must bind all parameters, even ones for which there was no corresponding argument. If there's an explicit default (`= default`) then it binds to that. Otherwise it binds to a new _instance_ of a default type corresponding to the parameter's sigil. (And a new collection instance is `.defined`, hence `@list.defined` is guaranteed to fail to do what is desired.)

**Comment by raiph** [comment]
The verbiage "Detecting Optional Arguments" in this issue's title is ambiguous.

It could reasonably be read as meaning something like this:
```
sub a (@list?) { say &?BLOCK.signature.params[0].optional }
a; # True
```
This issue seems to be discussing something else.

----

It could be this issue is more like "Detecting _Missing_ Arguments".

That could reasonably be read as meaning something like this:
```
sub a (|arguments (@list?)) { say arguments[0].defined }
a; # False
```

That might actually be desirable in some scenarios. So perhaps it ought be written up somewhere in the docs, and perhaps linked to or directly included in any doc written in response to this issue.

----

What this issue most seems to be about is sort of in between "Detecting Optional Arguments" and "Detecting Missing Arguments".

It seems to be about detecting whether the value that actually ends up being bound to a parameter, _after signature binding has completed and all parameters have been bound_, was or wasn't sourced from an argument.

And right now it seems to me that the best way to determine that is to add an explicit default that will be bound to a parameter if an argument isn't passed.

----

It may be appropriate to pick a default that cannot be passed, to avoid false positives -- thinking no argument was passed when one actually was:
```
sub a (@list = Positional) { say @list.defined }
a Positional; # False
```
If the above is a problem, then perhaps:
```
my role NoArg does Positional {}

sub a (@list = NoArg) is export { say @list.defined }
a; # False
```
The above ensures that no code that imports `a` can pass an argument that tricks `a` into a false positive.

**Comment by raiph** [comment]
Another case is slurpies, eg `*@foo`. In that case you'll need to analyze the capture (the list of arguments passed):
```
sub a ($foo, *@list (|args)) { say ?args }
a 42; # False
```

**Comment by librasteve** [comment]
Sorry for the slow response.

I agree that this Issue should be renamed "Detecting Missing Arguments".

So far we have seen two good ways to express this, so I amend my initial "Starter for 10" as follows. Hopefully we have a consensus on this and can proceed to make the change.

---
### Detecting Missing Arguments

Checking for the presence of missing arguments can be done like this for Scalars:
```
sub a($scalar = Nil) {  say $scalar.defined }
a      # False
```

For Positionals and Associatives, it is complicated by the fact that the empty List (), Array [] and Hash %{} are .defined.

A simple approach is to set an undefault value, such as a type class:
```
sub a (@list = Positional) { say @list.defined }
a       # False
```

Another approach, is to introspect the capture:
```
sub a (|arguments (@list?)) { say arguments[0].defined }
a; # False
```

The ```?``` optional marker (or ```is optional``` trait) has no effect when a default is specified.

Note: you can use the tests [] ~~ Empty (True) and {} ~~ Empty (True) to test for the presence of elements within Positional and Associative parameters respectively.

---

**Comment by raiph** [comment]
**TL;DR** Perhaps "Detecting Omitted Arguments"?

----

Thank you for renaming the issue title; I think that's a useful step; but to be clear I meant that as a thought experiment, a step in a direction of travel, not a final destination.

----

One of the underlying problems related to this issue that I had in mind is the linguistic (English) mental models folk have (as opposed to technical mental models), as reflected in the following SO exchange I had with the poster of a question:

I wrote:

> The "optional" in "optional parameter" doesn't mean the parameter is optional. Instead it means that passing an argument corresponding to that parameter is optional.

Their reply was:

> That's why you shouldn't say "optional parameter". Remember that people who have no idea what any of this is will read this and think that the docs mean what they say.

Their reply ironically couples an implicit demonstration of my implicit point (English is ambiguous) with what appears to be total ignorance of the relatively explicit point (that they had misinterpreted "optional parameter", because English is inherently ambiguous no matter how hard you might try to eliminate it).

Why else would they declare we should change the doc to match _their_ incorrect understanding? Did they really think we could arrive at some new phrasing that was less confusing about what is after all distilled in the parameter trait `is optional`?

Their reply seemed to reflect arrogance of the highest order, but still, what is one to do?

----

My current thinking is that "optional parameter" stays. (Unless someone comes up with a better term, and I'm not personally much interested in pursuing that. I think it will be challenging given that I can't imagine their being consensus that the `is optional` trait ought be deprecated, and I sure know I won't vote for that.)

But it does reflect subtle problems with ambiguous verbiage.

And while "detecting missing argument" is imo a significant improvement over "detecting optional arguments" (because what's being detected here is not the `is optional`), the switch to "detecting missing arguments" has its _own_ problems.

What's being detected here isn't some kind of mistake -- and to me "missing" connotes a problem ("who lost that argument?!?").

So what might we do as a next thought experiment (still not as a final destination, though perhaps closer or close enough)?

Perhaps "detecting _omitted_ arguments"?

**Comment by raiph** [comment]
**TL;DR** If the sketch you provided was just to further our understanding in this issue prior to thinking about a PR then that's great and it did the job for me. But "Starter for 10", "consensus", and "proceed to make the change", and the way you are writing things, makes me think you may have meant it as something akin to a PR draft. This comment is based on that interpretation and is push back about things I didn't like in that context. And then my next comment after this one will be my "Starter for 2c" about a PR draft.

> Checking for the presence of missing arguments can be done like this for Scalars:
```
sub a($scalar = Nil) {  say $scalar.defined }
a      # False
```

Some things I don't like about that:

* Starting with `$scalar = Nil`. I'd rather start with `$scalar?`, or `$scalar is optional`, or `($scalar is optional) { # \`$scalar is optional\` means the same as \`$foo?\`` or `($foo?) { # \`$foo\`? means the same as \`$foo is optional\``.

* The `= Nil` is redundant. Again, I get that it can serve as a way to establish a pattern, but I still don't like it.

* The use of `Nil` would fail to detect an omitted/missing argument if that argument is `Nil`.

* Worse, the `.defined` test will fail to detect an omitted/missing argument if that argument is not `.defined`.

> A simple approach is to set an undefault value, such as a type class:

I presume you're just sketching out how to think about things, rather than trying to nail down precise verbiage, right? Is "undefault" a typo, meant to be "default"? Or is touching on something else, like the fact that there's an implicit default default and what a user needs to do is set an explicit default? Is "type class" a slip, or because you don't like the existing term "type object", or something else?

Putting those questions aside, I don't like the formulation you sketch for roughly the same reasons as I the ones that meant I didn't like `= Nil` and `.defined`.

> Another approach, is to introspect the capture:

Written exactly like that could be a very abrupt introduction of "capture" if the reader hasn't encountered, or at least studied/understood them, before.

So I would suggest two things.

First, some preamble text, a succinct version of the following distinction which I don't have time right now to reduce to something sufficiently brief:

* Directly, laboriously, manually, introspecting the list of arguments passed, while thinking about the parameters they were to be bound to, and understanding the signature binding process. The strength and weakness of doing things this way is that one is dealing with the raw list of arguments passed, with complete disregard for the parameters, and how particular arguments ended up being paired up with particular parameters, other than one's own _thinking_ about, and understanding of, that process. Given that the whole notion of "omitted" is relative to how individual arguments get bound to individual parameters, direct introspection of arguments is a very fragile and error prone way to detect omission of any particular argument. This approach is only mentioned for completeness sake, and to motivate the _other_ (generally much more sensible) way to detect omission of arguments.

* Establishing appropriate _defaults_ for parameters (defaults are used if the signature binding process decides that a corresponding argument was omitted) and then appropriately introspecting and testing to see if a _parameter_ has ended up bound to the default after the signature binding process is complete. The defaults need to be appropriate: they need to be something that cannot be passed as an argument. , (Otherwise you can't tell the difference between an omitted argument and an argument that just happens to match the default.) And the test also needs to be appropriate: it needs to yield `True` iff the parameter is bound to the default, and not otherwise.
 
Secondly, I think the word/concept "capture" should be generally avoided as far as possible, perhaps with some text (eg "passed arguments") that links to the capture doc.

----
```
sub a (|arguments (@list?)) { say arguments[0].defined }
```
That was from what I wrote. It was terribly sloppy both because of writing `[0]` (when it won't contain anything) and relying on `.defined` then returning `False` (and `.defined` being the wrong thing to test in the first place because one will get false positives).

> The `?` optional marker (or `is optional` trait) has no effect when a default is specified.

Hmm. I'd express that as something more like setting a default means the same thing as `?` but sets an __explicit default_ rather than relying on the _implicit default default_.

> Note: you can use the tests [] ~~ Empty (True) and {} ~~ Empty (True) to test for the presence of elements within Positional and Associative parameters respectively.

Yeah, it might make sense to mention that in the verbiage preceding a general approach.

**Comment by raiph** [comment]
An initial installment on a potential PR...

# To detect an omitted argument

**Simple solution**

Give the parameter a default value. Test that the parameter ends up with the default value. If it did, assume the argument was omitted.

If the default value is one that can accidentally be passed, and you don't care about detecting that particular case, i.e. you are OK with assuming that the argument was _not_ passed even though in fact it _was_, then this simple solution is ideal (because it's simple). For example:
```
sub foo ($_ = 'NoName') {
    when 'NoName' { say 'no name passed' }
    say "'$_' passed";
 }
foo 'Fred'; # 'Fred' passed
foo; # 'no name passed'
```

If the default value is one that _cannot_ be passed, accidentally or deliberately, then this solution is again ideal, because it's simple _and bullet proof_. For example:
```
# SomeModule.rakumod;
unit module SomeModule;
my role ArgOmitted {}
sub foo ($_ = ArgOmitted) is export {
    when ArgOmitted { say 'no name passed' }
    say "'$_' passed";
 }

# main.raku
use lib '.';
use SomeModule;
foo 'Fred'; # 'Fred' passed
foo; # 'no name passed'
foo 'ArgOmitted'; # 'ArgOmitted' passed
foo ArgOmitted; # Compile time error
} 
```
**Complicated solutions**

If the above simple solution isn't appropriate, things get more complicated fast.

To recap, the simple solution won't be appropriate in the rather rare case that you want to set a default for a parameter that a caller might pass as a value, but you're just not willing to accept that you then can't distinguish between whether the parameter ended up with that default value due to an argument corresponding to that parameter having actually been passed with that value, or that argument having actually been omitted.

And to add to that recap: are you really sure you want to make your life a lot more fraught? :)

...

**Comment by coke** [comment]
To get this issue resolved, would really appreciate it if a PR can be submitted

**Comment by raiph** [comment]
**TL;DR** Some more verbiage that might belong in a **Simple Solution** section.

----

Raku allows routines to specify that one or more of its _parameters_ (the variables that are in the routine's _signature_ to be available for use in its body) are _optional_.

(NB. This doesn't mean that those _parameters_ may or may not exist. They exist, no matter what, end of story. What "optional" means is that the designer of the routine is declaring that corresponding _arguments_ may be omitted without that being an error.)

For example:
```
sub hi ($salutation = 'Hello', :$name = 'World') { say "$salutation $name" }
hi; # Hello World
hi 'Hello'; # Hello World
hi name => 'Gorgeous'; # Hello Gorgeous
```
Note that the first two invocations had the same result.
What if you wanted to detect whether `$salutation` was set from a passed argument, or the argument was omitted and the parameter was instead set to the default?

More generally, how do you detect whether an argument was passed in the call, or omitted from the call?

There are several ways to mark one or more parameters in a routine "optional":

* Explicitly giving it a default as above. In that case the explicit default is used if an argument corresponding to the parameter is omitted.

* Ending it with a `?` or ` is optional`. In that case an implicit default default is used if an argument corresponding to the parameter is omitted.

For example, consider this routine:
```
sub bye ($salutation = 'Bye cruel', $name?) { say $salutation, " ", $name // 'World' }
bye; # bye cruel World
bye 'see ya', 'seniōre'; # see ya seniōre
```
In the first call no arguments were passed, so `$name` was given the implicit default default for a `$foo` parameter, which is to say `(Any)`. That's why the `$name // 'World'` worked; it properly dealt with the fact that the `$name` parameter was optional and the corresponding argument was omitted.

**Comment by raiph** [comment]
**TL;DR** Is something like the simple solution verbiage I've posted above adequate as a PR? I think it's plausible that we should not be documenting the complicated case yet, and perhaps ever, other than perhaps briefly outlining options. This may, perhaps should, be coupled with posting a non-doc issue seeking relief through a new Raku(do) feature, in the hope that the new feature is implemented and properly puts this issue to bed, or at least that no one complains about the doc of this aspect for another year or five.

(The new feature would provide a way to introspect whether or not the signature binder bound a parameter to an argument from the call capture. I'd be personally loathe to suggest even filing such a request, and perhaps would downvote adding that complication to Raku(do) but that's just my current view which might change if folk demonstrate that the simple solution I've outlined, perhaps with some verbiage about more complicated solutions, is inadequate.)

**Comment by librasteve** [comment]
I'm closing this issue because it is diverging from a PR.

**Comment by librasteve** [comment]
I agree on the refinement to 'Omitted'.

I agree that ```Nil``` is too 'spiky' a default.

I agree that we are not proposing a change to the language at this point.

I see that there are a number of valid concerns implicit in your latest notes - to me the "bullet proof" one is the most material.

Since is past my bedtime, I will come back on your text specifics tomorrow.

**Comment by raiph** [comment]
Hi @librasteve,

**TL;DR** My "Final 2¢" response to your "Starter for 10". I'm off on a trip which may mean I barely do any Raku stuff for at least a couple weeks, maybe a month or more. Please feel free to use whatever you like in a real PR, or wait until I get back online. Thanks for pushing things forward and yet being patient while I figured out what feels to me like something close to a decent straw-dog "prePR" based on our exchange.

I've prepared the following "prePR" with the presumption it'll be inserted as a new section in the Signature literals page immediately after the existing [Optional and mandatory arguments](https://docs.raku.org/language/signatures#Optional_and_mandatory_arguments) section. It may need adjusting if it's used, at least in part, but lands elsewhere.

I've written it in GH markdown. If you need to extract it as such but don't have rights to edit my comment, @coke or someone else should be able to get that for you.

----

# Was an argument passed for a parameter?

Table showing tests of whether an argument was passed for a given parameter:

|Parameter type|Comment|Idiomatic test|
|-|-|-|
|[Slurpy](https://docs.raku.org/language/signatures#Slurpy_parameters), eg `*@array`|Don't test using `.defined`|Use eg `if @array`|
|Required, eg `$foo`|Can't be omitted|(not applicable)|
|Optional, eg `@bar = default`|Pick a suitable default¹|eg `if @bar =:= NotPassed`|

Table showing results of testing if a parameter is set to its default value:

|Default¹|Result of test|Argument was passed?|
|-|-|-|
|`NotPassed`¹|Parameter is _not_ set to its default|Yes|
|`NotPassed`¹|Parameter _is_ set to its default|No|
|`'Passable'`¹|Parameter is _not_ set to its default|Yes|
|`'Passable'`¹|Parameter _is_ set to its default|Maybe²|

Footnotes:

¹ A _passable_ default is a value that could actually be passed by a caller, thus partially defeating the ability to tell if an argument was passed via a default check. A _not-passed_ default is a value that _cannot_ be passed by a caller, thus ensuring that a check of it is guaranteed to determine whether or not an argument was passed.²

Example of a passable default value:
```
sub pass-check (Str $_ = 'Passable') is export {
    when 'Passable' { say 'Unsure if arg was passed' }
    say "$_ passed";
 }
pass-check 'Another string'; # Another string passed
pass-check;                  # Unsure if arg was passed
pass-check 'Passable';       # Unsure if arg was passed
```

² If you want a guarantee that you can tell for sure whether or not an argument corresponding to an "optional" parameter was passed, then use a not-passed default.¹

Example of a not-passed default value:
```
module not-passed-default {
  my class NotPassed is Str {}
  sub pass-check(Str $_ = NotPassed) is export {
      when NotPassed { say 'No arg passed, guaranteed' }
      say "$_ passed";
  }
 }
import not-passed-default;
pass-check 'Some string'; # Some string passed
pass-check;               # No arg passed, guaranteed
#pass-check NotPassed;    # Compile error. Code can't pass the module's NotPassed.
```

**Comment by librasteve** [comment]
@raiph - this is good progress - thanks for providing a polished PR-ready proposal

I have two related points:

1. I gather that the inclusion of a module / export structure is to wrap the ```NotPassed``` value to avoid accidental use elsewhere. I think that this idea is quite complex and, noting that you employ ```=:=``` to check for identity, I think that there is a way to avoid that.
2. I see that you make a [valid] distinction between ```Passable``` and ```NonPassed```, I hope that we can use identity to avoid that.

----

Here's a modified "prePR" with the presumption it'll be inserted as a new section in the Signature literals page immediately after the existing [Optional and mandatory arguments](https://docs.raku.org/language/signatures#Optional_and_mandatory_arguments) section.

----

# Was an argument passed for a parameter?

Table showing checks of whether an argument was passed for a given parameter:

|Parameter kind|Example|Comment|Check for no arg passed|
|-|-|-|-|
|[Slurpy](https://docs.raku.org/language/signatures#Slurpy_parameters)|`*@array`|Don't check using `.defined`|`if not @array`|
|Required|`$foo`|Can't be omitted|(not applicable)|
|Optional|`@bar = default`¹|Pick, and check, a suitable default¹|`if @bar === default {...}`¹|

A parameter with a default is always _optional_, so there is no need to mark it with a `?` or the `is optional` trait.

¹A suitable default is an Object that has a distinct identity, as may be checked by the [`WHICH`](https://docs.raku.org/type/Mu#method_WHICH) method. Then you can use the `===` [value identity operator](https://docs.raku.org/language/operators#infix_===) in the body of the routine to check whether this exact default was bound to the parameter.

In raku, Numeric, Stringy and type object items _are_ identical so they are not suitable for this purpose. We need to pick a default that creates a new object and to declare it as a `constant`.

Table showing suitable default values for untyped parameters:

|Sigil|Type|Default²|Check|
|-|-|-|-|
|`$`|Scalar|`constant AnyAt² = Any.new`|`if $x === AnyAt {...}`|
|`@`|Positional|`constant PositionalAt² = Positional.new`|`if @x === PositionalAt {...}`|
|`%`|Associative|`constant AssociativeAt² = Associative.new`|`if %x === AssociativeAt {...}`|

²This example uses names like `TypeclassAt` to reflect that the `.WHICH` test returns an object of type `ObjAt`, you are free to make up you own names.

Example with a positional parameter:
```
constant PositionalAt = Positional.new;

sub a (@list = PositionalAt ) { say @list =:= PositionalAt }
a;              #True 
a [1,2,3];      #False
```

Example with some scalar parameters:
```
constant AnyAt = Any.new; 

sub b ($x=AnyAt, :$y=AnyAt) { say $x === AnyAt, $y === AnyAt }
b 1;            #FalseTrue 
b 1,:2y;        #FalseFalse
```

If your parameters are typed, then the [type smileys](https://docs.raku.org/language/glossary#Type_smiley) can be used  with [`multi`](https://docs.raku.org/language/functions#Multi-dispatch)s like this:

```
multi sub c (Int:U $z)  { say 'Undefined' }
multi sub c (Int:D $z)  { say 'Defined'   }   
multi sub c (Int   $z?) { say 'Omitted'   }   
c;              #Omitted 
c (Int);        #Undefined 
c 42;           #Defined
``` 

----

**Comment by raiph** [comment]
@librasteve 

> inclusion of a module / export structure is to wrap the `NotPassed` value to avoid accidental use elsewhere.

It's to make the `NotPassed` variant do what it says on the tin. As I'll explain below, your alternate solution doesn't achieve that.

(`NotPassed` isn't just about "accidental use". It also eliminates _abuse_, ie malicious code trying to trick code that uses the `'Passable'` variant. I'm presuming that, if someone really wants to bother with something as esoteric as checking if arguments got past, it's worth them thinking about protecting against abuse. Or not; they might instead want to allow for "useful abuse", ie _benign_ code trying to trick code that uses the `'Passable'` variant, at the risk of malign abuse -- which they might ensure is defended against if indeed it's even possible. Etc.)

But perhaps you knew all that anyway; I just wanted to be clear.

----

> I think that [`NotPassed`] is quite complex

Yes, very.

In my "first installment" of this "PR" I almost balked at including the concept behind the `NotPassed` variant at all. That said, it is still conceptually way simpler than some other approaches I thought / suspect the doc might end up discussing.

So I totally get the reaction of "Oh, the children! Can we reduce this complexity?"

There are definitely ways to substantially alter the readers experience of complexity. But for this "Final prePR" I decided to switch to MegaWATT Light Bulb writing.¹ I may have ended up blinding unsuspecting readers! Whether that risk is taken just for this "PR" or an actual PR will be up to you (unless you choose to wait until I get back to this stuff).

That said, I do ask that you and others seriously ponder to what degree it is appropriate to try ratchet back the complexity in this particular case given that it's unavoidable (if someone wants what `NotPassed` does, even if it's done slightly differently), and mull this conundrum for a good week to see how things feel after that mulling.¹

----

> you employ `=:=` to check for identity, I think that there is a way to avoid that.

Maybe that was _one new thing too many_, but to be clear, `=:=` is dead simple (iiuic).

Aiui it's the simplest of Raku's equality operators, indeed the simplest any PL could have. Aiui it compares the underlying value in RAM and quite literally can't get any simpler.

I may be wrong about that, but I still understand it to be just about the simplest equality operator Raku can support _technically_ (even if, perhaps, not conceptually, because Raku is generally so high level).

But few Rakoons use it, so it becomes one more _new_ thing to learn. "Oh, Raku has a bog standard basic low level equality operator? I didn't know that!". Each new thing to learn can be the straw that breaks the camel's back. (Even if they're a lambdacamel. If one ends up blind, even the best MegaWATT journey ends up sucking.)

----

Let's distinguish between two things.

First, why is `NotPassed` so darn complex overall, and why is it presented right out of the gate, in such a terse form? For that, see footnote 1.

Second, why `=:=`, given it's likely a new operator to a reader? Again, see footnote 1, and discussion of the MegaWATT journey.

So I grant that the whole thing seems super complex. And I get that `=:=` adds yet another thing.

But I also would like you to reflect on why a Rakoon who is not ready to learn advanced / new things, would have a problem with a section of the doc that's really about an advanced topic?

(I get that one might not think that detecting omitted arguments is an advanced topic, but in most ways it really it is, for reasons I've discussed in earlier comments.)

> I see that you make a [valid] distinction between Passable and NonPassed, I hope that we can use identity to avoid that.

Do you mean avoid _the distinction_?

If so, I think that reflects something I didn't explicitly mention. Consciously. (But perhaps inadvisably? Perhaps I should have mentioned "it" in introducing the "PR"? Or perhaps in the verbiage of the "PR" itself?)

What I mean by "it" is one of the two motivations for covering `'Passable'`.

One motivation was was as a stepping stone, or fallback, or starting point, for readers understanding `NotPassed`, especially if they find `NotPassed` very complicated.

But the other is that sometimes -- often in fact -- someone wants to set the default to some value that has nothing to do with whether an argument was passed or not. And you can't really set _two_ defaults (one for testing whether an argument was passed, and the other to set the default value used for other purposes). For example, the default might be part of automatically generated API doc. So `'Passable'` is going to be a common case.

But then you're inevitably subject to accidents and abuse that trick the `'Passable'` variant.

So what to do about that? Well, use the `NotPassed` variant.

(But then you've lost the ability to use the default in the usual way. So what to do about that? Well, use a `Passable` variant. But then, well, then you'll ask on IRC and someone will point out that you can directly introspect the capture, but that's incredibly fraught and fragile. I really didn't want that in the doc if we could help it! If we reach the point there's pressure to put that in the doc in this section, it may be time to open an issue for a new Raku(do) feature...)

----

If you _didn't_ mean "avoid the distinction" but rather avoid `=:=`, then I think you may have picked the wrong thing to tackle. Yes, it's a new thing, and might be the straw that broke the camel's back, but is it the right thing to focus on?

First, I think it would be better to remove one of the bigger pieces of burden. Perhaps my _up-fronting_ of the `NotPassed` variant, rather than the `'Passable'` variant (in a reversal of my treatment in the earlier "first installment" comment about what became this "PR"), needs to be rereversed.

And/or, perhaps you can come up with some technique that's simpler than a lexically hidden new type. I tried and failed to think of one. Now it's your turn to try!

But also I'm confused about (your view, or perhaps mine, about) `=:=`:

> Also, when writing my examples, I noticed that `===` is what we want, not `=:=` ;-)

Why are you saying `===` is wanted and `=:=` isn't?

As I understand it `===` is for comparing a "value type", which is to say values that might not _actually_ be the same, but _compare_ the same nonetheless. I was thinking that that might open things up to spoofing.

(Though perhaps not. The values do have to be the same type, and that may mean `=:=` and `===` are ultimately about the same. Though I still think `===` is a ton more complex. Though it may be more familiar...)

Aiui, `=:=` just compares two values without interpreting them. Thus:
```
say 1 =:= 1; # True -- integers are interned, so the two `1`s are the same object reference.
say 'a' =:= 'a'; # True -- small strings are interned, so the two `'a'`s are again the same object.
say 'a' x 1000 =:= 'a' x 1000; # False -- aiui long strings and/or roped strings aren't interned
my $a = 42;
my $b := $a;
say $a =:= $b; # True -- `$a` and `$b` are aliases, and Scalars evaluate to their content
```
Note the difference:
```
say 'a' x 1000 === 'a' x 1000; # True -- they're actually different but value compare the same
```

----

> In raku, `Numeric`, `Stringy` and type object items are identical so they are not suitable for this purpose.

I'm not following your logic here.

I _do_ get the logic of the `NotPassed` variant seeming to be very complicated. Because it is. So I get there's a problem: it's complicated. What I mean is I don't follow your proposed approach to an alternate _solution_.

The key thing for the `NotPassed` variant is to introduce a unique value (object, anything Raku code can set as a default, and use in a comparison) that _cannot be passed_.

Consider your code:
```
constant PositionalAt = Positional.new;

sub a (@list = PositionalAt) { say @list =:= PositionalAt }
a;              #True 
a [1,2,3];      #False
```
One can abuse this:
```
constant PositionalAt = Positional.new;

sub a (@list = PositionalAt) { say @list =:= PositionalAt }
a;              #True 
a [1,2,3];      #False
a PositionalAt; # True
```
In the later case, the test says no argument was passed. But one was: `PositionalAt`.

That was the entire point of `NotPassed`. You cannot spoof it.

----

So I would say that's a red herring you pursued.

That said, if so, we should perhaps fall back to _why_ you pursued it.

And that's because `NotPassed` is complex.

I reacted to all of it getting so complex by being bloody determined to see if I could keep the doc section short, accept the inherent complexity, reduced to its bare minimum, and whack readers with it all in a single fairly short sharp section.

After all, I reasoned, just about no Rakoon should ever want to do this (explicitly check whether an argument was passed), and in the very rare case they did, most should be fine with `'Passable'`, and in the even rarer case they weren't, they should be fine with the `NotPassed` variant, and in the incredibly rare case that even that isn't good enough, well, I was thinking we'd cross that bridge if we ever arrive at it.

----

BTW, in writing this monster comment I've realized I lost another comment I intended to post that outlined why Raku is so good in this respect, and how just about no Rakoon would ever need to even read this doc section, but ended up distilling it into the first table, which implicitly makes the point.

To make it more explicit:

* Raku automatically takes care of managing arguments -> parameters so most Rakoons don't need to ever worry about "was an argument passed?".

* If it's a slurpy, it gets 0 or more arguments. Just write `if @foo` if you care to test if any arguments were passed.

* If it's a required non-slurpy, it'll get an argument, period.

* Even if it's an optional, there should pretty much never be a reason to check if an argument was passed. If it was, the parameter is set to it. If it wasn't the parameter is set to some suitable default. Don't worry about it, and life is fine.

* In the incredibly rare case you _must_ know if an argument was passed, and it's a non-slurpy optional, then here's what you can do that's sane: Set default; compare parameter with default.

(And we hopefully avoid mentioning checking captures. Though I had decided we could perhaps reasonably include that. One would have to iterate a signature and capture in parallel, first pairing required positionals and nameds, and then optional positionals, and then optional nameds after first sorting both signature and capture keys. Or something like that.)

----

A final thing is you reintroducing code that varies between sigils. I think that just makes it seem like you _had_ to introduce that extra stuff because it won't make sense without it, and I think that will make the reader experience _worse_ because they won't realize that you didn't need to add that extra sigil specific stuff.

And to be clear, I don't think you did need to add it. I would expect that any Rakoon reading that section and deciding they wanted to understand it would have all the knowledge they need to figure out how to pick a default (which was distilled into `NotPassed`, which you'll see `is Str`, thus laying the intellectual ground for a `NotPassed does Positional` etc) and then test it via comparison.

That said, I get the impression you just don't see things as I do on that front, and it's time to wrap this Novella.

Peace and love to all, I gotta run and this should be my last word on this topic for the next few weeks at least!

# Footnote

¹ **First**, my overall approach for this so-called "Final prePR" was to _maximize_ the payoff of a "decision" to somewhat reverse what I'd done in my earlier "first installment". In that earlier installment I led the explanation with the `'Passable'` variant, which then motivated the `NotPassed` variant. In this one I somewhat reversed that and/or presented it as one narrative.

Why the reversal / one combined narrative? Well let's assume it doesn't matter, because in a sense it doesn't. Instead what matters is that (it feels to me like) we ought document both.

(If we're going to document this aspect at all; frankly I think this is all pretty advanced stuff that doesn't fit well in doc I'd expose to newbies if I had my druthers. But this issue is predicated on requesting a new piece of doc, and I've chosen not to pushback on the very request itself and instead acquiesce on coming up with a PR to see whether it is ultimately rejected or evolved and merged.)

So then the question becomes, how should they be presented? Focus on `'Passable'`, which then motivates `NotPassed`, or vice-versa, or kind of combined as I've done them?

I thought I should take the opportunity to work it through somewhat reversed.

And once I'd done that, i.e. once I'd decided to introduce the fundamentally more complicated concept right up front, I wanted that to pay off for the reader. My hope was that, given that the _concept_ of `NotPassed` was much more complicated (introduction of a new type, lexically scoping it, trying to convey the, er, import of the lexical scoping, where by "import" I mean "importance") I'd at least be able to keep the _explanation_ of the concept as succinct as possible while still giving enough for a pay off I'll describe in a moment.

And I decided I'd be most likely to generate that payoff if I could guide readers who cared to understand to read code that then hit them up the side the head with not one but four aspects at once (new type, lexical scope, modules, `=:=`).

Then you and other Rakoons can decide what's best, if indeed y'all agree the overall approach (focus on setting the default to some special value, and testing for that special value) is the way to go.

**Second**, there's the payoff I'm talking about, the thing I was attempting to maximize. What am I getting at? How could hitting anyone upside the head with lots of concepts simultaneously, none of which seem to have anything to do with figuring out whether an argument was passed, ever be the right thing to do?

Well, I'm not going to go so far as to say it *is* the right thing to do in this case. One has to be _very_ careful about giving folk a cognitive stretch goal when they didn't ask for it and just wanted a quick answer to their simple question ("Was an argument passed?"). I'm just saying it _might_ be -- because it's important that one takes advantage of teachable moments.

If someone is motivated to read a section titled "Was an argument passed?", then either they're reading it casually, so can just move on if they don't really care, or they want an answer to that specific question, which should motivate them to try figure out what on earth is going on, either by thinking, or by searching around and reading, or by asking around.

So my goal was that pretty much anyone who tried to understand what the doc meant, and why the heck a module turned up of all things, would be in the middle of a "teachable moment" journey that could be, and hopefully is in this "PR", a MegaWATT Light Bulb Trip. That is, they would end up going through something like the following stages:

1. MegaWAT. "WAT? I thought I was just going to read about how to know if an argument had been passed! And I get instructions on how to build a bloody rocket?"

2. "WTF is going on?" "I should be able to work this out!!!" "In like 10 seconds flat! "Grrr!"

3. If they know enough Raku, they figure it out. It might take a few minutes before the penny drops, or it might take longer or less time. But they figure it out, and it makes enough sense that they're fairly confident why it is as it is. (They might even think the doc could make that clearer. And they might be right. But there can be negative costs of making the doc more complete, in the mistaken belief that every WTF should be eliminated. But if they're motivated, they can create a PR.) So they figure it out, and a light bulb did eventually light up. It may have lit dimly or brightly, but it did, and then the WTF!?! feeling subsides, evolving into an "Ah, hmm, OK." or perhaps a "Yeah, hmm, that's fairly clean I guess", or maybe "Ah, cool, I get it, and can explain that to someone else".

4. Or, if they _don't_ know enough Raku, they end up saying "WTF?" or asking WTF is going on. And some Rakoon knows it enough to explain it to them and to other Rakoons or other readers.

5. All of this capitalizes on a teachable moment. In this case it's more complicated than I'd prefer. I introduce declaring a new type (`class foo is bar {}` declares a new type). And lexical scoping (what `{ ... }` enforces). _And_ modules/imports (controlling how identifiers do or don't escape from a lexical scope). I'm _glad_ that this seemed (almost?) too much to you; it did to me. I pined after some way to simplify the identity aspect, like you did. I fully expect my `NotPassed` make many Rakoons, especially newbies, glaze over. But then why are they reading this section? If they don't really need to know they'll just move on, perhaps remarking on it to someone, and perhaps benefiting from point 4 above. Conversely, if they really _do_ want or need to know then they are presumably already into fairly advanced Raku notions, and should hopefully be motivated and thus both willing to go on the journey of learning whichever of those three aspects (new type, lexical scope, module/import) they don't already know, and then enjoy the one, two, three, or four Light Bulb switch-on moments I packed into `NotPassed`.

**Comment by librasteve** [comment]
> Do you mean avoid the distinction?   
 
yes

> On `=:=` vs. `===`

I looked at this again in a vacuum and I think that, for the examples I proposed, in  this distinction is moot

```
constant e = Any.new;
constant f = Any.new;

say e =:= f;    #False       \   ... either of these works and both apply `.WHICH` semantics for this sort of object
say e === f;    #False     /
```

Then I checked your latest and TIL about `value internment`. Again, in my examples, this is moot since I apply the comparison only to non-interned objects.

So, I agree that `=:=` is the best comparison operator for our purpose.

> In raku, Numeric, Stringy and type object items are identical so they are not suitable for this purpose.

Thanks for your feedback, I will change this phrasing (see below)

> On `NotPassed`

I think your example of possible _abuse_ is at the crux of our disagreement. I repeat it here:

```
constant PositionalAt = Positional.new;

sub a (@list = PositionalAt) { say @list =:= PositionalAt }
a;              #True 
a [1,2,3];      #False
a PositionalAt; # True    <== abuse
```

The question I am aiming to answer is "how can I detect that an argument was omitted when the sub was called". And not "how can I prevent malicious code from ever passing an argument without my being able to check for this". I am not saying that the latter would never be needed, just that this _is_ an advanced topic that we do not need in the general docs.

My answer is somewhat resistant to this vector of abuse in that [if we use `my constant`] we can confine "access" to the information of the actual ID of our selected PositionalAt to the scope and then have a fairly robust way to prevent external code (eg. from a web frontend, API, file, external module) from being able to make the same PositionalAt.

I would be happy to see your `NotPassed` example in an SO question - but I do think it goes too far in terms of advancing the topic.

----

Updated (and hopfully final) PrePR in the next comment....

**Comment by librasteve** [comment]
Updated (and hopfully final) PrePR...

----

# Was an argument passed for a parameter?

Table showing checks of whether an argument was passed for a given parameter:

|Parameter kind|Example|Comment|Check for no arg passed|
|-|-|-|-|
|[Slurpy](https://docs.raku.org/language/signatures#Slurpy_parameters)|`*@array`|Don't check using `.defined`|`if not @array`|
|Required|`$foo`|Can't be omitted|(not applicable)|
|Optional|`@bar = default`¹|Pick, and check, a suitable default¹|`if @bar === default {...}`¹|

A parameter with a default is always _optional_, so there is no need to mark it with a `?` or the `is optional` trait.

Footnotes:

¹A suitable default is an Object that has a distinct identity, as may be checked by the [`WHICH`](https://docs.raku.org/type/Mu#method_WHICH) method. Then you can use the `=:=` [container identity operator](https://docs.raku.org/language/operators#infix_=:=) in the body of the routine to check whether this exact default was bound to the parameter.

Example with a positional parameter:
```
my constant PositionalAt = Positional.new;

sub a (@list = PositionalAt) { say @list =:= PositionalAt }
a;              #True 
a [1,2,3];      #False
```

Example with some scalar parameters:
```
my constant AnyAt = Any.new; 

sub b ($x=AnyAt, :$y=AnyAt) { say $x =:= AnyAt, $y =:= AnyAt }
b 1;            #FalseTrue 
b 1,:2y;        #FalseFalse
```

If your parameters are typed, then the [type smileys](https://docs.raku.org/language/glossary#Type_smiley) can be used  with [`multi`](https://docs.raku.org/language/functions#Multi-dispatch)s like this:

```
multi sub c (Int:U $z)  { say 'Undefined' }
multi sub c (Int:D $z)  { say 'Defined'   }   
multi sub c (Int   $z?) { say 'Omitted'   }   
c;              #Omitted 
c (Int);        #Undefined 
c 42;           #Defined
``` 

The examples use names like `TypeclassAt` to reflect that the `.WHICH` test returns an object of type `ObjAt`, you are free to make up you own names.

**Comment by librasteve** [comment]
@gfldex ... I have reopened the issue so we don't ignore you suggestion

in defence of the current "final prePR" [fppr] proposal:

- a single method to detect omitted arguments across all sigil cases is desirable
- the fppr doesn't initially introduce types since we need an answer for untyped parameters anyway
- the solution is generally easier for typed parameters, as can be seen from the fppr `multi` examples
- the `multi` example is close to yours - however `::T` type captures are quite an advanced topic
